### PR TITLE
WIP: Null Check improvements

### PR DESCRIPTION
**Pull request by tribals**
For #116.

### CODE REVIEW

**Comment by tribals** [comment]
In fact, it is very annoying to write such kind of tests. These can be written in another way:

```python
from unittest import TestCase

from javalang.parser import Parser
from javalang.tokenizer import tokenize

from aibolit.patterns.null_check.null_check import NullCheck


class TestNullCheck(TestCase):
    def test_null_check_comparison_result_assignment(self):I
        snippet = 'boolean i = z == null;'

        node = Parser(tokenize(snippet)).parse_block_statement()

        self.assertEqual(NullCheck()._traverse_node(node), [1])
```

Here is two questions:
1. Is such style of tests accepted? Or it is better to always write a "real" Java class?
2. It requires us to call private method. Is it accepted as well?

**Comment by 0crat** [comment]
This pull request [#156](https://github.com/yegor256/aibolit/issues/156) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/b20ab37f-7c0a-4949-953b-a743a9ad34ea); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by acheshkov** [comment]
@lyriccoder @yegor256 please say yours two cents.

**Comment by acheshkov** [comment]
The point makes sense. But I'm not sure that we can do the same for any pattern. Actually we can, but some patterns deal with the structure that span more than one line and hence you will be  going to put this code as text to your test case.   


```python
snippet = 'here will be multi line snippet;'
```

**Comment by tribals** [comment]
This is completely not a problem since Python has multi-line strings:

```python
snippet = """\
public void MyMethod(int x) {
  if (x == null) {}
}
"""
```

The only thing need to be done is calling right method of `javalang.parser.Parser` - in this example, it must be `.parse_member_declaration()`.

In fact, I propose to write *every* Java example code, even complete classes, inside test itself. These examples seem to be small, so there is no problem in keeping them inside test.

**Comment by tribals** [comment]
Since nobody reacted, I assume this style is accepted.

**Comment by acheshkov** [comment]
Let's give it a shot and see

**Comment by fanifieiev** [comment]
@tribals @acheshkov I would actually suggest having `value ` method that would accept 
`javalang.ast.Node` but not `filename`, so that there is no need to call `NullCheck()._traverse_node(node)`

`_traverse_node(node)` looks like a private/internal method.

So, I would suggest:
```
class NullCheck(object):
    def value(self, javalang.ast.Node):
       ....
```
So, if someone needs to get the source code from a filename:
`result = NullCheck().value(AST(filename))`

If someone need to get the source code from snippet:
`result = NullCheck().value(_parser(snippet).parse_block_statement())`
I would also suggest having a separate class:
```
class BlockStatement:
    def __init__(self, snippet):
        self.snippet = snippet

    def value():
        return _parser(snippet).parse_block_statement()
```

and use the above as:
`result = NullCheck().value(
        BlockStatement(
             snippet
        ).value()
 )`

**Comment by tribals** [comment]
> @tribals @acheshkov I would actually suggest having `value ` method that would accept
> `javalang.ast.Node` but not `filename`, so that there is no need to call `NullCheck()._traverse_node(node)`

I'm absolutely agree with you! But this change deserves it's own issue. I would suggest to adapt to static type checking first, please see: #130 and #165. After doing so, we can introduce a real interface for patterns, rather an ephemeral one (also know as "Duck type"). The next step would be to enable type-check in CI pipeline, so no new patterns could be accepted without conforming to that `Pattern` interface.

**Comment by fanifieiev** [comment]
> > @tribals @acheshkov I would actually suggest having `value ` method that would accept
> > `javalang.ast.Node` but not `filename`, so that there is no need to call `NullCheck()._traverse_node(node)`
> 
> I'm absolutely agree with you! But this change deserves it's own issue. I would suggest to adapt to static type checking first, please see: #130 and #165. After doing so, we can introduce a real interface for patterns, rather than an ephemeral one (also know as "Duck type"). The next step would be to enable type-check in CI pipeline, so no new patterns could be accepted without conforming to that `Pattern` interface.

@tribals Actually I created an issue long ago to create interface 
https://github.com/yegor256/aibolit/issues/123

**Comment by tribals** [comment]
@fanifieiev Excellent! Haven't seen it, though.

Well, seems like everything already in place in order to make patterns more solid at it's behavior. We need only to close issues. :slightly_smiling_face:

**Comment by 0crat** [comment]
@fanifieiev[/z](https://www.0crat.com/u/fanifieiev) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

File: test/patterns/null_check/test_null_check.py [hunk 1]
```diff
@@ -20,23 +20,68 @@
...
-        file = Path(self.cur_file_dir, 'NullCheck.java')
+        snippet = """\
+        if (this.z == null) { // here!
+            throw new RuntimeException("oops");
+        }
+        """
 
-        self.assertEqual(self.testClass.value(file), [5])
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_block_statement()), [1]
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:43]: @tribals I think it is better to:
`NullCheck()._traverse_node(_parser(snippet).parse_block_statement()), [1]` as each test method is a dedicated test unit.
It is also not good to have this  `_traverse_node` method calling here.

PS: Actually each pattern should accept the `snippet `or `filename `or etc via constructor, but not the `value `method, but this is out of scope of this PR.

File: test/patterns/null_check/test_null_check.py [hunk 2]
```diff
@@ -20,23 +20,68 @@
...
+        snippet = """\
+        public NullCheck(String z) {
+            if (z == null) { // here!
+                throw new RuntimeException("oops");
+            }
+        }
+        """
+
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_member_declaration()),
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:56]: @tribals Same as above

File: test/patterns/null_check/test_null_check.py [hunk 3]
```diff
@@ -20,23 +20,68 @@
...
+
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_block_statement()), [1]
+        )
+
+    def test_null_check_ternary(self):
+        snippet = 'luckyName == null ? luckyName : "No lucky name found";'
+
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_block_statement()), [1]
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:71]: @tribals Same as above

File: test/patterns/null_check/test_null_check.py [hunk 4]
```diff
@@ -20,23 +20,68 @@
...
+
+    def test_null_check_not_equal_comparison(self):
+        snippet = """\
+        if (this.z != null) { // here!
+            throw new RuntimeException("oops");
+        }
+        """
+
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_block_statement()), [1]
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:82]: @tribals Same as above

File: test/patterns/null_check/test_null_check.py [hunk 5]
```diff
@@ -20,23 +20,68 @@
...
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_member_declaration()),
+            [],
+        )
+
+    def test_null_check_comparison_result_assignment(self):
+        snippet = "boolean i = z == null;"
+
+        self.assertEqual(
+            self.pattern._traverse_node(_parser(snippet).parse_block_statement()), [1]
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:64]: @tribals Same as above

File: test/patterns/null_check/test_null_check.py [hunk 6]
```diff
@@ -20,23 +20,68 @@
...
+from javalang.tokenizer import tokenize
+
 from aibolit.patterns.null_check.null_check import NullCheck
 
 
 class TestNullCheck(TestCase):
-    cur_file_dir = Path(os.path.realpath(__file__)).parent
-    testClass = NullCheck()
+    def setUp(self):
+        self.pattern = NullCheck()
```
Comment (fanifieiev) [inline test/patterns/null_check/test_null_check.py:33]: @tribals According to below comment, it is better to remove this part

File: test/patterns/null_check/test_null_check.py (hunk 1 above)
Comment (tribals) [reply test/patterns/null_check/test_null_check.py:43]: There is no need to create `NullCheck` explicitly in each test method since `setUp` is called once per every test method. Please, see: https://docs.python.org/3/library/unittest.html#organizing-tests.

I agree, though, that it is not clear from source code itself.

> It is also not good to have this `_traverse_node` method calling here.

I absolutely agree. A method which name is prefixed with underscore considered a private one, and is not allowed to call directly. But, since there is only one possible way to pass already parsed AST to pattern now, I suggest to leave it as is until we will fix all patterns.

**Comment by tribals** [comment]
@fanifieiev Please, take a look at issue itself: #116. There was some pattern which are not implemented yet in this PR. Namely:

```
assertThrows(NullPointerException.class, () -> wrapperSum(null, 2));
Optional op2 = Optional.ofNullable(null);
Objects.requireNonNull(bar, "bar must not be null");
```

Unfortunatelly, I don't understand, what exactly should be considered a pattern? I requested clarification from reporter, but there is still none.

Could we separate them to individual issue and accept work done for now?

**Comment by fanifieiev** [comment]
> @fanifieiev Please, take a look at issue itself: #116. There was some patterns which are not implemented yet in this PR. Namely:
> 
> ```
> assertThrows(NullPointerException.class, () -> wrapperSum(null, 2));
> Optional op2 = Optional.ofNullable(null);
> Objects.requireNonNull(bar, "bar must not be null");
> ```
> 
> Unfortunatelly, I don't understand, what exactly should be considered a pattern? I requested clarification from reporter, but there is still none.
> 
> Could we separate them to individual issue and accept work done for now? It already covers the wast majority of null checks.

@tribals Absolutely, you can leave a `@todo` with the explanation of what to do. Please find an example of Todo [here](https://github.com/yegor256/aibolit/blob/a1c8969d7fa897ba8f263675eb89cb1dbb21365a/aibolit/patterns/var_middle/var_middle.py#L72) if you do not know how to create a todo

File: test/patterns/null_check/test_null_check.py (hunk 1 above)
Comment (fanifieiev) [reply test/patterns/null_check/test_null_check.py:43]: > There is no need to create `NullCheck` explicitly in each test method since `setUp` is called once per every test method. Thus, `self.pattern` is a new one in each test. Please, see: https://docs.python.org/3/library/unittest.html#organizing-tests.
> 
> I agree, though, that it is not clear from source code itself.
> 
> > It is also not good to have this `_traverse_node` method calling here.
> 
> I absolutely agree. A method which name is prefixed with underscore considered a private one, and is not allowed to call directly. But, since there is only one possible way to pass already parsed AST to pattern now, I suggest to leave it as is until we will fix all patterns.

@tribals I know that `setup` does it for each test method, but I personally(not only me) find that setup method is very confusing and brings more confusions. 
I think it is better to create pattern object in each test method  separately.
@acheshkov WDYT ?

Comment (tribals) [reply test/patterns/null_check/test_null_check.py:43]: Well, personally I find the same about setup methods, too. So, it is simple to me to follow you advice.

**Comment by tribals** [comment]
@fanifieiev Please, take a look.

**Comment by fanifieiev** [comment]
@tribals Please create a @todo for this case https://github.com/yegor256/aibolit/pull/156#issuecomment-610612129

**Comment by tribals** [comment]
@fanifieiev Done

**Comment by acheshkov** [comment]
@rultor merge
